### PR TITLE
Throw a binder exception for invalid date strings.

### LINKED ISSUE #1335: Core Dump On Invalid Date String
## Summary
We are hard failing (core dump) when parsing a bad date string. We are correctly identifying the problem, but are throwing an exception that causes the server to crash. We should instead catch the exception then report the error message back to the client.

## Steps to Reproduce

```SQL
noisepage=# CREATE TABLE types_test (col DATE);
CREATE TABLE
noisepage=# INSERT INTO types_test (col) VALUES ('2020-1-1'),('20-1-1');
INSERT 0 2
noisepage=# SELECT * FROM types_test WHERE col < '0000-01-01';
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```

**Server**
```
terminate called after throwing an instance of 'noisepage::ConversionException'
  what():  0-1-1 is not a valid date

#1  0x00007ffff78d3859 in __GI_abort () at abort.c:79
#2  0x00007ffff7ca8951 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff7cb447c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff7cb44e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff7cb4799 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00005555578ef265 in noisepage::execution::sql::Date::FromYMD (year=0, month=1, day=1) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/execution/sql/runtime_types.cpp:233
#7  0x00005555578eef3c in noisepage::execution::sql::Date::FromString (str=0x7fff58007f54 "0000-01-01", len=10) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/execution/sql/runtime_types.cpp:227
#8  0x000055555765d4c9 in noisepage::execution::sql::Date::FromString (str="0000-01-01") at ../src/include/execution/sql/runtime_types.h:138
#9  0x0000555557aea9d8 in noisepage::binder::BinderUtil::CheckAndTryPromoteType (value=..., desired_type=noisepage::type::TypeId::DATE) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/binder_util.cpp:69
#10 0x0000555557ae58fb in noisepage::binder::BindNodeVisitor::Visit (this=0x7fffee7fb310, expr=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/bind_node_visitor.cpp:666
#11 0x000055555754ce61 in noisepage::parser::ConstantValueExpression::Accept (this=0x7fff58007ed0, v=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/parser/expression/constant_value_expression.cpp:347
#12 0x000055555753d8b8 in noisepage::parser::AbstractExpression::AcceptChildren (this=0x7fff58005e10, v=...) at ../src/include/parser/expression/abstract_expression.h:202
#13 0x0000555557aec2a9 in noisepage::binder::SqlNodeVisitor::Visit (this=0x7fffee7fb310, expr=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/sql_node_visitor.cpp:30
#14 0x0000555557ae56eb in noisepage::binder::BindNodeVisitor::Visit (this=0x7fffee7fb310, expr=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/bind_node_visitor.cpp:647
#15 0x000055555754a4ef in noisepage::parser::ComparisonExpression::Accept (this=0x7fff58005e10, v=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/parser/expression/comparison_expression.cpp:24
#16 0x0000555557ae3a84 in noisepage::binder::BindNodeVisitor::Visit (this=0x7fffee7fb310, node=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/bind_node_visitor.cpp:468
#17 0x000055555748bdb1 in noisepage::parser::SelectStatement::Accept (this=0x7fff58006080, v=...) at ../src/include/parser/select_statement.h:349
#18 0x0000555557adf9a9 in noisepage::binder::BindNodeVisitor::BindNameToNode (this=0x7fffee7fb310, parse_result=..., parameters=..., desired_parameter_types=...)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/binder/bind_node_visitor.cpp:61
#19 0x0000555557267f57 in noisepage::trafficcop::TrafficCop::BindQuery (this=0x555558e2ddf0, connection_ctx=..., statement=..., parameters=...)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/traffic_cop/traffic_cop.cpp:311
#20 0x0000555557653959 in noisepage::network::SimpleQueryCommand::Exec (this=0x7fff580092f0, interpreter=..., out=..., t_cop=..., connection=...)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/postgres/postgres_network_commands.cpp:168
#21 0x0000555557658abb in noisepage::network::PostgresProtocolInterpreter::Process (this=0x7fff64004e90, in=..., out=..., t_cop=..., context=...)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/postgres/postgres_protocol_interpreter.cpp:45
#22 0x000055555768dc45 in noisepage::network::ConnectionHandle::Process (this=0x7fff58000ef0) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:182
#23 0x000055555769622f in noisepage::network::ConnectionHandleStateMachineTransition::Process (handle=...) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:76
#24 0x000055555768d6d1 in noisepage::network::ConnectionHandle::StateMachine::Accept (this=0x7fff58000f10, action=noisepage::network::Transition::WAKEUP, handle=...)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:129
#25 0x000055555768db20 in noisepage::network::ConnectionHandle::HandleEvent (this=0x7fff58000ef0, fd=126, flags=2) at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:169
#26 0x000055555768de3c in noisepage::network::ConnectionHandle::<lambda(int, int16_t, void*)>::operator()(int, int16_t, void *) const (__closure=0x0, fd=126, flags=2, arg=0x7fff58000ef0)
    at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:217
#27 0x000055555768de72 in noisepage::network::ConnectionHandle::<lambda(int, int16_t, void*)>::_FUN(int, int16_t, void *) () at /home/wanshenl/Documents/RareTerrierPhotos/noisepage/src/network/connection_handle.cpp:218
#28 0x00007ffff7ece13f in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#29 0x00007ffff7ece87f in event_base_loop () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
#30 0x000055555767a177 in noisepage::EventUtil::EventBaseDispatch (base=0x7fff64000ff0) at ../src/include/common/event_util.h:100
#31 0x000055555767a1b4 in noisepage::common::NotifiableTask::EventLoop (this=0x7fff64000ee0) at ../src/include/common/notifiable_task.h:151
#32 0x000055555767a1f2 in noisepage::common::NotifiableTask::RunTask (this=0x7fff64000ee0) at ../src/include/common/notifiable_task.h:158
#33 0x000055555769e44a in noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}::operator()() const (this=0x555558d22330)
    at ../src/include/common/dedicated_thread_registry.h:82
#34 0x00005555576c06db in std::__invoke_impl<void, noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}>(std::__invoke_other, noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}&&) (__f=...) at /usr/include/c++/9/bits/invoke.h:60
#35 0x00005555576c0625 in std::__invoke<noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}>(noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}&&) (__fn=...) at /usr/include/c++/9/bits/invoke.h:95
#36 0x00005555576c0596 in std::thread::_Invoker<std::tuple<noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x7fff64001d58) at /usr/include/c++/9/thread:244
#37 0x00005555576c050e in std::thread::_Invoker<std::tuple<noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}> >::operator()()
    (this=0x7fff64001d58) at /usr/include/c++/9/thread:251
#38 0x00005555576c03c2 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<noisepage::common::DedicatedThreadRegistry::RegisterDedicatedThread<noisepage::network::ConnectionHandlerTask, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory> >(noisepage::common::DedicatedThreadOwner*, unsigned int, noisepage::common::ManagedPointer<noisepage::network::ConnectionHandleFactory>)::{lambda()#1}> > >::_M_run() (this=0x7fff64001d50) at /usr/include/c++/9/thread:195
```

### Expected Behavior
Once again, Postgres does it correctly:

```SQL
pavlo=# CREATE TABLE types_test (col DATE);
CREATE TABLE
pavlo=# INSERT INTO types_test (col) VALUES ('2020-1-1'),('0020-1-1');
INSERT 0 2
pavlo=# SELECT * FROM types_test WHERE col < '0000-01-01';
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: SELECT * FROM types_test WHERE col < '0000-01-01';
                                             ^
LOCATION:  DateTimeParseError, datetime.c:3731
Time: 0.333 ms
```

### PR DESCRIPTION
**Pull request by amogkam**
Small PR to fix issue described in https://github.com/cmu-db/noisepage/issues/1335.

Previously, when parsing an invalid date string, the server would crash
```
terminate called after throwing an instance of 'noisepage::ConversionException'
  what():  0-1-1 is not a valid date
Aborted (core dumped)
```
and the client would show:
```
noisepage=# SELECT * FROM types_test WHERE col < '0000-01-01';
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```

With this PR, the exception is caught and propagated to the client:
```
noisepage=# SELECT * FROM types_test WHERE col < '0000-01-01';
ERROR:  0-1-1 is not a valid date
```
and the server doesn't crash.

@apavlo @lmwnshn would be great if you guys could take a look! Please let me know if there's a better way to handle this error or a better message to use.

Also, note that this treats all string to date conversion errors as `ERRCODE_INVALID_DATETIME_FORMAT`, though Postgres also has a separate [` ERRCODE_DATETIME_VALUE_OUT_OF_RANGE`](https://github.com/cmu-db/noisepage/blob/master/src/include/common/error/error_code.h#L118) and [`ERRCODE_DATETIME_FIELD_OVERFLOW`](https://github.com/cmu-db/noisepage/blob/master/src/include/common/error/error_code.h#L117) which would each need to be handled separately I believe.

Closes #1335

### CODE REVIEW

File: src/binder/binder_util.cpp [hunk 1]
```diff
@@ -66,9 +66,13 @@ void BinderUtil::CheckAndTryPromoteType(const common::ManagedPointer<parser::Con
...
           case type::TypeId::DATE: {
-            auto parsed_date = execution::sql::Date::FromString(str_view);
-            value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
-            break;
+            try {
+              auto parsed_date = execution::sql::Date::FromString(str_view);
+              value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
+              break;
+            } catch (ConversionException &exception) {
+              throw BINDER_EXCEPTION(exception.what(), common::ErrorCode::ERRCODE_INVALID_DATETIME_FORMAT);
```
Comment (lmwnshn) [inline src/binder/binder_util.cpp:74]: Thanks for the PR!

1. I tried a bunch of queries and it seems that they generally prefer error 22008 ERRCODE_DATETIME_FIELD_OVERFLOW for whatever reason, could you switch to that error code?

```
postgres=# create table types_test (a date);
CREATE TABLE
postgres=# select * from types_test  where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
LOCATION:  DateTimeParseError, datetime.c:3731
postgres=# drop table types_test;
DROP TABLE

postgres=# create table types_test (a timestamp);
CREATE TABLE
postgres=# select * from types_test  where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test  where a < '0000-01-01';
                                            ^
LOCATION:  DateTimeParseError, datetime.c:3731
```

2. Could you add a similar thing for timestamps right below? I believe the timestamp conversion code may also throw ConversionError.

Thanks!

**Review by lmwnshn — CHANGES_REQUESTED** [review]
Minor changes requested.

File: src/binder/binder_util.cpp (hunk 1 above)
Comment (amogkam) [reply src/binder/binder_util.cpp]: Thanks for the review @lmwnshn!

I modified it to use error code 22008 and added another try-catch for timestamps.

With regards to the error code, it's dependent on the input
```
postgres=# create table types_test (a date);
CREATE TABLE
postgres=# select * from types_test where a < '0000-01-01';
ERROR:  date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test where a < '0000-01-01';
                                           ^
postgres=# \errverbose
ERROR:  22008: date/time field value out of range: "0000-01-01"
LINE 1: select * from types_test where a < '0000-01-01';
                                           ^
LOCATION:  DateTimeParseError, datetime.c:3731
postgres=# select * from types_test where a < 'dsijfdi';
ERROR:  invalid input syntax for type date: "dsijfdi"
LINE 1: select * from types_test where a < 'dsijfdi';
                                           ^
postgres=# \errverbose
ERROR:  22007: invalid input syntax for type date: "dsijfdi"
LINE 1: select * from types_test where a < 'dsijfdi';
                                           ^
LOCATION:  DateTimeParseError, datetime.c:3758
```

I believe we'd have to parse the input to get the correct error code, but I think this is out of scope for this PR. 

I actually think using error code 22007 as a catch-all would be the best option here until we can differentiate between them. I'm happy to switch this back if you agree.

Comment (lmwnshn) [reply src/binder/binder_util.cpp]: That sounds reasonable to me. Feel free to switch it back, could you please add a comment alongside it (basically what you already wrote)? Thanks!

Comment (amogkam) [reply src/binder/binder_util.cpp:76]: Great, I just updated it!

**Review by lmwnshn — APPROVED** [review]
I'm not sure if the linter will let you write `TODO:` without a name attached. If it does, LGTM, otherwise you might need to `TODO(amogkam):` Thanks!

File: src/binder/binder_util.cpp [hunk 2]
```diff
@@ -66,14 +66,26 @@ void BinderUtil::CheckAndTryPromoteType(const common::ManagedPointer<parser::Con
...
-            auto parsed_date = execution::sql::Date::FromString(str_view);
-            value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
-            break;
+            try {
+              auto parsed_date = execution::sql::Date::FromString(str_view);
+              value->SetValue(type::TypeId::DATE, execution::sql::DateVal(parsed_date));
+              break;
+            } catch (ConversionException &exception) {
+              // For now, treat all conversion errors as 22007.
+              // TODO: Differentiate between 22007 and 22008. See comments in PR #1462.
```
Comment (amogkam) [inline src/binder/binder_util.cpp:75]:
```suggestion
              // TODO(amogkam): Differentiate between 22007 and 22008. See comments in PR #1462.
```

File: src/binder/binder_util.cpp [hunk 3]
```diff
@@ -66,14 +66,26 @@ void BinderUtil::CheckAndTryPromoteType(const common::ManagedPointer<parser::Con
...
-            auto parsed_timestamp = execution::sql::Timestamp::FromString(str_view);
-            value->SetValue(type::TypeId::TIMESTAMP, execution::sql::TimestampVal(parsed_timestamp));
-            break;
+            try {
+              auto parsed_timestamp = execution::sql::Timestamp::FromString(str_view);
+              value->SetValue(type::TypeId::TIMESTAMP, execution::sql::TimestampVal(parsed_timestamp));
+              break;
+            } catch (ConversionException &exception) {
+              // For now, treat all conversion errors as 22007.
+              // TODO: Differentiate between 22007 and 22008. See comments in PR #1462.
```
Comment (amogkam) [inline src/binder/binder_util.cpp:86]:
```suggestion
              // TODO(amogkam): Differentiate between 22007 and 22008. See comments in PR #1462.
```

**Comment by amogkam** [comment]
@lmwnshn should we merge this?

**Comment by lmwnshn** [comment]
Merging, thanks for the PR!

**Pull request merged**